tests/functional/test_s3.py

```diff
@@ -1431,6 +1438,55 @@ def test_content_sha256_not_set_if_config_value_is_false(self):
         sha_header = sent_headers.get("x-amz-content-sha256")
         self.assertEqual(sha_header, b"UNSIGNED-PAYLOAD")
 
+    def test_content_sha256_set_if_config_value_not_set_put_object(self):
+        # The default behavior matches payload_signing_enabled=False. For
+        # operations where the `Content-MD5` is present this means that
+        # `x-amz-content-sha256` is present but not set.
+        config = Config(signature_version="s3v4")
+        self.client = self.session.create_client(
+            "s3", self.region, config=config
+        )
```

Comment:
Did we intentionally set `s3v4` or was this carryover from other tests? If it's not proving something specific, maybe we want to leave this unconfigured to make sure s3 is choosing the right signer.

Reply:
This was a carryover from the other tests. 

At a high level: 
* The only test that fails before the fix in this PR and passes afterwards is in `test_basic_outpost_arn` where this PR adds three lines (719-721). This is in commit 384f88641.
* All changes to the `TestS3SigV4` class are tangential and improve test coverage and add inline comments because those tests misled us during the investigation of this bug. These are in commit 2f2b1acc2.